packages/graphiql/src/components/GraphiQL.tsx

```diff
@@ -773,6 +881,134 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
               </button>
             )}
           </div>
+          {this.props.tabs ? (
+            <Tabs>
+              {this.state.tabs.tabs.map((tab, index) => (
+                <Tab
+                  key={tab.id}
+                  isActive={index === this.state.tabs.activeTabIndex}
+                  title={tab.title}
+                  isCloseable={this.state.tabs.tabs.length > 1}
+                  onSelect={() => {
+                    this.handleStopQuery();
+                    this.setState(state => {
+                      const oldActiveTabIndex = state.tabs.activeTabIndex;
+                      const tabs = state.tabs.tabs.map(
+                        (currentTab, tabIndex) => {
+                          if (tabIndex !== oldActiveTabIndex) {
+                            return currentTab;
+                          }
+
+                          return {
+                            ...currentTab,
+                            query: state.query,
+                            variables: state.variables,
+                            operationName: state.operationName,
+                            headers: state.headers,
+                            response: state.response,
+                            hash: idFromTabContents({
+                              query: state.query,
+                              variables: state.variables,
+                              headers: state.headers,
+                            }),
+                          };
+                        },
+                      );
+
+                      const newActiveTab = this.state.tabs.tabs[index];
+
+                      return {
+                        ...state,
+                        query: newActiveTab.query,
+                        variables: newActiveTab.variables,
+                        operationName: newActiveTab.operationName,
+                        headers: newActiveTab.headers,
+                        response: newActiveTab.response,
+                        tabs: { ...state.tabs, tabs, activeTabIndex: index },
+                      };
+                    }, this._persistTabsState);
+                  }}
+                  onClose={() => {
+                    if (this.state.tabs.activeTabIndex === index) {
+                      this.handleStopQuery();
+                    }
+                    this.setState(state => {
```

Comment:
as per the earlier request, I'd like to see if we can have some reusable functions external to this file for these state handlers.  Really don't want to increase complexity in this style, and most of these could be handled as pure reducer-like standalone functions within the arguments that can be later migrated into hooks

Reply:
Perhaps even the `Tabs` component could do this, and handle the tab state internally? and accept the required handlers as props for now?

Reply:
The tabs component is not able of handling all the states as the state is also affected by various other event handlers. E.g. `handleEditQuery`.

I agree that we can refactor the whole state into a reducer in a follow-up pull request. I don't see why doing that would block advancing on this tabs feature.

In https://github.com/graphql/graphiql/pull/2197/commits/1c6ad348a35a04b3a1256760e8b28afc6f8abcc6 I moved the state update logic into reducer-like functions.

In general, I am not a big fan of separating state strictly from components and having to jump around too much for simple changes.

Reply:
I guess it's a common pattern for monolithic react library components to be one huge class component, essentially. People complain about the cognitive overhead, but you're correct, we can find ways to simplify it later. In the RFC effort two years ago we reduced this file to under 1000 lines, and moved a ton of the state into context. But we can still do that in later iterations, sure!